File: ios/Runner/MediaPipePlugin.swift
import Flutter
import UIKit

class MediaPipePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var helper = FaceLandmarkerHelper()
    private var frameTimestamp = 0

    static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "com.skinapp/mediapipe",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.skinapp/mediapipe_events",
            binaryMessenger: registrar.messenger()
        )
        let instance = MediaPipePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            helper.onResult = { [weak self] (landmarks, w, h) in
                DispatchQueue.main.async {
                    if landmarks.isEmpty {
                        self?.eventSink?(["detected": false])
                    } else {
                        self?.eventSink?([
                            "detected": true,
                            "landmarks": landmarks,
                            "imageWidth": w,
                            "imageHeight": h
                        ])
                    }
                }
            }
            helper.onError = { [weak self] error in
                self?.eventSink?(FlutterError(code: "MEDIAPIPE_ERROR", message: error, details: nil))
            }
            helper.setup()
            result(true)

        case "processFrame":
            guard let args = call.arguments as? [String: Any],
                  let bytes = args["bytes"] as? FlutterStandardTypedData,
                  let isFront = args["isFrontCamera"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing args", details: nil))
                return
            }
            frameTimestamp += 33
            helper.detectAsync(imageData: bytes.data, timestamp: frameTimestamp, isFrontCamera: isFront)
            result(nil)

        case "dispose":
            helper.close()
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/mycompany/dermascanai/MediaPipePlugin.kt
package com.mycompany.dermascanai

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.os.Handler
import android.os.Looper
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream

class MediaPipePlugin(private val context: Context) :
    MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private var faceLandmarkerHelper: FaceLandmarkerHelper? = null
    private var eventSink: EventChannel.EventSink? = null
    private var frameTimestamp = 0L

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                setupLandmarker()
                result.success(true)
            }
            "processFrame" -> {
                val bytes = call.argument<ByteArray>("bytes") ?: return result.error("NO_DATA", "No frame bytes", null)
                val isFront = call.argument<Boolean>("isFrontCamera") ?: true
                val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                if (bitmap == null) {
                    result.error("DECODE_FAIL", "Could not decode frame bytes", null)
                    return
                }
                frameTimestamp += 33
                faceLandmarkerHelper?.detectAsync(bitmap, frameTimestamp, isFront)
                result.success(null)
            }
            "processYuvFrame" -> {
                val yuvBytes = call.argument<ByteArray>("yuvBytes") ?: return result.error("NO_DATA", "No YUV bytes", null)
                val width = call.argument<Int>("width") ?: return result.error("NO_DATA", "No width", null)
                val height = call.argument<Int>("height") ?: return result.error("NO_DATA", "No height", null)
                val isFront = call.argument<Boolean>("isFrontCamera") ?: true

                val jpegBytes = try {
                    convertYuvToJpeg(yuvBytes, width, height)
                } catch (e: Exception) {
                    result.error("YUV_CONVERT_FAIL", "YUV→JPEG conversion failed: ${e.message}", null)
                    return
                }

                if (jpegBytes == null) {
                    result.error("YUV_CONVERT_FAIL", "YUV→JPEG returned null", null)
                    return
                }

                val bitmap = BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
                if (bitmap == null) {
                    result.error("DECODE_FAIL", "Could not decode converted JPEG", null)
                    return
                }

                frameTimestamp += 33
                faceLandmarkerHelper?.detectAsync(bitmap, frameTimestamp, isFront)
                result.success(jpegBytes)
            }
            "dispose" -> {
                faceLandmarkerHelper?.close()
                faceLandmarkerHelper = null
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    private fun setupLandmarker() {
        faceLandmarkerHelper = FaceLandmarkerHelper(
            context = context,
            onResult = { landmarkerResult, width, height ->
                Handler(Looper.getMainLooper()).post {
                    sendLandmarksToFlutter(landmarkerResult, width, height)
                }
            },
            onError = { error ->
                Handler(Looper.getMainLooper()).post {
                    eventSink?.error("MEDIAPIPE_ERROR", error, null)
                }
            }
        )
        faceLandmarkerHelper?.setup()
    }

    /**
     * Convert NV21 / YUV420 byte array to JPEG using Android's native [YuvImage].
     */
    private fun convertYuvToJpeg(nv21Bytes: ByteArray, width: Int, height: Int): ByteArray? {
        val yuvImage = YuvImage(nv21Bytes, ImageFormat.NV21, width, height, null)
        val out = ByteArrayOutputStream()
        yuvImage.compressToJpeg(Rect(0, 0, width, height), 85, out)
        return out.toByteArray()
    }

    private fun sendLandmarksToFlutter(result: FaceLandmarkerResult, width: Int, height: Int) {
        if (result.faceLandmarks().isEmpty()) {
            eventSink?.success(mapOf("detected" to false))
            return
        }

        val landmarks = result.faceLandmarks()[0].map { lm ->
            mapOf("x" to lm.x().toDouble(), "y" to lm.y().toDouble(), "z" to lm.z().toDouble())
        }

        eventSink?.success(
            mapOf(
                "detected" to true,
                "landmarks" to landmarks,
                "imageWidth" to width,
                "imageHeight" to height
            )
        )
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
